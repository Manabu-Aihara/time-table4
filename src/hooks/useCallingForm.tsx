import { useState, useRef, useCallback, useEffect, PropsWithChildren } from "react";
import { chakra } from "@chakra-ui/system";

import { TimelineEventProps, EventFormProps } from "../lib/TimelineType";

export const useCallingEditForm = ({onShowFormView, targetEvent}: EventFormProps) => {

  /**
   * Issue summary & progress
   */
  const [showModal, setShowModal] = useState(false);

  // TypeScriptでReactのイベントにどう型指定するか
  // https://komari.co.jp/blog/10724/
  const handleOuterFormBubbling = (e: React.MouseEvent<HTMLDivElement>) => {
    if(!(e.target instanceof HTMLButtonElement)){
      return;
    }
    setShowModal(false);
  }

  const countRef = useRef<number | undefined>();
  const handleSelectEvent = useCallback(
    (callingEvent: TimelineEventProps, e?: React.SyntheticEvent<HTMLElement, Event>) =>
  {
    // console.log(`Event motion: ${e.type}`);
    onShowFormView(callingEvent);
    countRef.current = undefined;
    console.log('切り替わりました Handle: ', countRef.current);
    setShowModal(true);
  }, []);

	const divRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    // console.log('Effect通りました', eventFlag);
    divRef.current?.scrollIntoView({behavior: 'smooth'});
  }, [targetEvent]);

  const closeInputForm = () => {
    setShowModal(false);
  }

  const EditForm: React.FC<PropsWithChildren> = ({children}) => {
    return (
      <chakra.div flexShrink="0" scrollSnapAlign="start"
        onClick={handleOuterFormBubbling}>
          {children}
      </chakra.div>
    )
  }

  const modal = {showModal, closeInputForm};
  return {handleSelectEvent, EditForm, modal};
}
