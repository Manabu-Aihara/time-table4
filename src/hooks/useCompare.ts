import { useRef, useEffect } from 'react';
import { TimelineEventProps } from "../lib/TimelineType";

//実際に使うhook
export const useChangeDebugger = (props: TimelineEventProps[]) => {
  const previousValue = usePrevious(props);

  const getChange = previousValue && getChanges(previousValue, props);

  if (getChange) {
    getChange.forEach((change) => console.log(change?.valueOf.call(previousValue)));
  // return getChange?.valueOf.call(previousValue);
  }
};

export const usePrevious = (props: TimelineEventProps[]) => {
  const previousValue = useRef<TimelineEventProps[] | null>(null);

  useEffect(() => {
    previousValue.current = props;
  });

  return previousValue.current;
};

function getChanges(previousValue: TimelineEventProps[],
  currentValue: TimelineEventProps[]) {
  if (
    typeof previousValue === "object" &&
    previousValue !== null &&
    typeof currentValue === "object" &&
    currentValue !== null
  ) {
    return Object.entries(currentValue).reduce<unknown[]>((acc, cur) => {
      const [key, value] = cur;
      const oldValue = previousValue[Number(key)];

      if (value !== oldValue) {
        acc.push({
          name: key,
          previousValue: oldValue,
          currentValue: value,
        });
      }

      return acc;
    }, []);
  }

  if (previousValue !== currentValue) {
    return [{ previousValue, currentValue }];
  }

  return [];
}