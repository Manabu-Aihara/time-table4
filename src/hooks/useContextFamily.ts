import { useContext } from "react";

import { EventsStateContext, EventsDispatchContext } from "../components/templates/EventsParent";
import { AuthStateContext, AuthDispatchContext } from "../components/templates/AuthParent";

export function useEventsState() {
  const state = useContext(EventsStateContext);
  if (!state) throw new Error('EventProvider not found');
  return state;
}
  
function useEventsDispatch() {
  const dispatch = useContext(EventsDispatchContext);
  if (!dispatch) throw new Error('EventProvider not found');
  return dispatch;
}

export function useAuthContext() {
  const auth = useContext(AuthStateContext);
  if (!auth) throw new Error('AuthProvider not found');
  return auth;
}

function useAuthDispatch() {
  const auth = useContext(AuthDispatchContext);
  if (!auth) throw new Error('AuthProvider not found');
  return auth;
}
