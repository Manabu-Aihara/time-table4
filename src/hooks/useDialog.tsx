import { ComponentProps, useCallback, useState } from "react";

import { Dialog as Component } from "../components/organisms/Dialog";

type UseDialogProp = Omit<
  ComponentProps<typeof Component>,
  "isOpen" | "onClose" | "rootElement"
>;

type Result = {
  open: () => void;
  Dialog: React.FC<UseDialogProp>;
  close: () => void;
};

export const useDialog = (): Result => {
  const [isOpen, setOpen] = useState<boolean>(false);

  const open = useCallback((): void => {
    console.log('Open the Dialog');
    setOpen(true);
  }, []);

  const close = useCallback((): void => {
    setOpen(false);
  }, []);

  const Dialog: React.FC<UseDialogProp> = useCallback(
    (prop: UseDialogProp): React.ReactNode => {
      return <div><Component isOpen={isOpen} onClose={close} {...prop} /></div>;
    }, [isOpen]);

  return { open, close, Dialog };
};
