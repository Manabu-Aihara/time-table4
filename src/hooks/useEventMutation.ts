import { useMutation, useQueryClient } from "@tanstack/react-query";

import basicAxios from "../lib/AuthInfo";
import { PickDate, TimelineEventProps } from "../lib/TimelineType";
import { eventKeys, useEventCache } from "../resources/cache";

// ⑦ mutationFn
// export const mutation = {
//   createTodo: (request: PostTodoRequest) => {
//     return todoApi.postTodo(request);
//   },
// };

export const useCreateMutation = () => {
  const eventCache = useEventCache();

  return useMutation({
    mutationFn: (timelineEvent: TimelineEventProps) =>
      basicAxios.post(`/event/add`, timelineEvent),
    onSuccess: () => {
      return eventCache.invalidateList();
    }
  });
}

export const useDeleteMutation = (targetId: number | string) => {
  const eventCache = useEventCache();

  return useMutation({
    mutationFn: () =>
      basicAxios.delete(`/event/remove/${targetId}`),
    onError: (error) => {
      console.log(`error!: ${error}`);
    },
    onSuccess: () => {
      console.log('サクセス通ってます');
      return eventCache.invalidateList();
    }
  });
}

export const useUpdateEventMutation = (targetId: number | string) => {
  const queryClient = useQueryClient();
  const eventCache = useEventCache();

  return useMutation({
    mutationFn: (timelineEvent: TimelineEventProps) =>
      basicAxios.post(`/event/update/${targetId}`, timelineEvent),
    onMutate: (timelineEvent) => {
      queryClient.setQueryData(eventKeys.detail(targetId), timelineEvent);
    },
    onError: (error) => console.log(`error!: ${error}`),
    onSuccess: () => {
      eventCache.invalidateList();
    },
  });
}

export const useUpdateDateListMutation = (targetIds: string[]) => {
  const queryClient = useQueryClient();
  const eventCache = useEventCache();

  return useMutation({
    mutationFn: (timeBeltArray: TimelineEventProps[]) => 
      basicAxios.post(`/date/update`, {
        data: timeBeltArray
      }),
    onMutate: (newDate) => {
      const prevEvents = queryClient.getQueryData(eventKeys.list());
      queryClient.setQueryData(eventKeys.dateList(targetIds),
        newDate
      );
      return { prevEvents };
    },
    // 一回引っかかって、ここで終了してしまう
    // onError: (error, variables) => {
    //   console.log(`error!: ${error}`);
    //   console.log(`variables: ${JSON.stringify(variables)}, context: prevEventsと同じよ`)
    // },
    onSettled: (data, error) => {
      console.log(`Mutation data: ${data}`);
      console.log(`Mutation error: ${error}`);
      eventCache.invalidateList();
    }
  });
}

const useUpdateDateMutation = (targetId: number | string) => {
  const queryClient = useQueryClient();
  const eventCache = useEventCache();

  return useMutation({
    mutationFn: (timeBelt: PickDate) => 
      basicAxios.post(`/date/update/${targetId}`, timeBelt),
    onMutate: (newDate) => {
      const prevEvents = queryClient.getQueryData(eventKeys.detail(targetId));
      queryClient.setQueryData(eventKeys.date(targetId),
        newDate
      );
      return { prevEvents };
    },
    onError: (error, variables, context) => {
      console.log(`error!: ${error}`);
      console.log(`variables: ${variables}, context: ${JSON.stringify(context)}`)
    },
    onSuccess: () => {
      eventCache.invalidateDetail(targetId);
    }
  });
}
