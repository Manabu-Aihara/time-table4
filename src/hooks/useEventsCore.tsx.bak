import { ComponentProps, useState, useCallback, useReducer } from 'react';
import { Event } from 'react-big-calendar';

type Events = Event[];
type State = {
  events: Events;
  isShow: boolean;
}

const initialStateFactory = (initialState: Partial<State>): State => ({
  events: [],
  isShow: false,
  ...initialState
});

// 汎用性高い* Reducer *
const reducer = <T extends { [key: string]: any }>(
  state: T,
  action: { type: keyof T; value: any }
) => {
  return { ...state, [action.type]: action.value } as T;
};

// type InputField<C extends keyof React.ComponentProps<'input'>> = {
//   state: C
// }
type InputField = ComponentProps<'input'>;

type DialogForm = {
  inputFields: InputField[]
}

export const useFormPages = <T extends { [key: string]: any }>(
  initialValues: T,
  form: DialogForm
) => {
  const [formValues, dispatch] = useReducer(reducer, initialValues);

  const [isShow, setIsShow] = useState(false);
  const show = useCallback(() => setIsShow(true), []);

  const showDialog = form.inputFields.map((field, i) => {
    const key = field.key as string;

    const handleChange = useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        dispatch({
          type: "SHOW", value: e.currentTarget.value });
      },
      [dispatch]
    );

    return (
      <div>
      <form onSubmit={onSubmit}>
        <input
          {...field}
          placeholder="やることを入力してくださいー"
          onChange={handleChange}
        />
        <button>追加</button>
      </form>
    </div>
    );
  })
}
