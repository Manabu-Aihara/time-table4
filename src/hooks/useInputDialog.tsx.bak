import { useState, FormEvent, ReactNode } from 'react';

import { useEventsDispatch, useEventsState } from '../lib/UseContext';
import { eventsReducer } from '../components/EventsParent';

type InputElementProps = React.ComponentProps<'input'>;

type DialogProps = {
  isOpen: boolean;
  children: ReactNode;
  close: () => void;
};

export const useInputDialog = () => {
  const [title, setTitle] = useState<string>('');
  // const [eventItem, setEventItem] = useState<EventItem>(null)
  const currentState = useEventsState();
  const dispatch = useEventsDispatch();

  const InputDialog = ({children}: {children: ReactNode}) => {

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      // このコンポーネントで onChange 発火時に必ず実行したい振る舞いを書く
      // const {name, value} = e.target;
      setTitle(e.target.value);
      // console.log(`ここにも注目：${title}`);
    };

    const onSubmit = (e: FormEvent) => {
      e.preventDefault();
      dispatch({
        type: 'CREATE',
        payload: {title: title}
      });
      // console.log(`ここ注目：${JSON.stringify(currentState)}`);
      const nextStage = eventsReducer(currentState, {type: 'CREATE', payload: {title: title}});
      console.log(`ここ注目：${JSON.stringify(nextStage)}`);
    };

  // const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
	// 	e.preventDefault();
	// 	fetch('http://127.0.0.1:8000/todo/add', {
	// 		method: 'POST',
	// 		headers: {
	// 			'Access-Control-Allow-Origin': '*',
	// 			// mode: 'no-cors',
	// 			// Accept: 'application/json',
	// 			'Content-Type': 'application/json'
	// 		},
	// 		body: JSON.stringify({
	// 			// todo
	// 			summary: todo.summary,
	// 			owner: todo.owner,
	// 			done: todo.done
	// 		})
	// 	})
	// 	.then(res => res.json());
	// 	// .then(json => console.log(json))
	// 	// .catch(err => console.log(err));
	// }

  return (
      <div>
        {children}
          <input
            // {...inputProps}
            placeholder="やることを入力してくださいー"
            onChange={handleChange}
          />
          <button onClick={onSubmit}>追加</button>
      </div>
    );
  }

  return InputDialog;
}