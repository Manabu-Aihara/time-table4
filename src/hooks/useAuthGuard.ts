import { useAuthContext } from "../hooks/useContextFamily";
import { useAuthQuery } from '../resources/queries';
import { AuthInfoProp } from '../lib/TimelineType';

export const useAuthInfo = () => {
  const authContext = useAuthContext();
  const tokenContext = authContext.type === 'token' ? authContext.accessToken : undefined;  
  
  const { data } = useAuthQuery(tokenContext!);

  const strData = JSON.stringify(data?.data);
  // パターン 1
  const objValue = JSON.parse(strData);
  // console.log(`Json type: ${typeof objValue.staff_id}, ${typeof objValue.group_id}`);
  const guard: AuthInfoProp = {
    authId: objValue.staff_id, code: objValue.group_id, group: objValue.group_name, type: 'auth'
  } as const;

  return guard;
}
