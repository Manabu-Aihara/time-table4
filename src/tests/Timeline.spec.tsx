import { act, render, waitFor, renderHook } from '@testing-library/react';
import { composeStories } from '@storybook/react';
import { expect, vi } from 'vitest';

import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import * as stories from '../stories/Timeline.stories';
import React from 'react';
import { useEventsState } from '../hooks/useContextFamily';

const createWrapper = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        // ✅ turns retries off
        retry: false,
      },
    },
  })

  return ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={queryClient}>
      {children}
    </QueryClientProvider>
  )
}

// it("my first test", async () => {
//   const { result } = renderHook(() => useEventsState(), {
//     wrapper: createWrapper()
//   });
// });
// const useGroupUsersQueryMock = vi.hoisted(() => 
//   vi.fn(() => ({
//     data: [{ id: 1, title: 'group 1' }, { id: 2, title: 'group 2' }]
//   }))
// )
// vi.mock('./resources/queries', () => ({
//   useGroupUsersQueryMock,
// }));

describe('MyHorizonTimeline component', () =>{
  const { Standard } = composeStories(stories);
  it('タイムラインの表示', async () => {
    const { container, getByText } = render(Standard());
    // const { result } = renderHook(() => useEventsState(), {
    //   wrapper: createWrapper()
    // });
    const titleElement = getByText('マイタイムライン');
    // expect(container).toMatchSnapshot();
    // const currentData = await waitFor(() => result.current.slice(-1)[0]);
    // expect(currentData).toBeDefined();
    await act(() => {
      Standard.play?.({ canvasElement: container});
    });
  });
  const { ModuleMock } = composeStories(stories);
  it('モックの動作確認', async () => {
    const { container } = render(ModuleMock());
    expect(container).toMatchSnapshot();
    expect(container).toBeDefined();
  })
});
