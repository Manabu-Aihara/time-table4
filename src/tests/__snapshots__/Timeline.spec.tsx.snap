// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyHorizonTimeline component > モックの動作確認 1`] = `
<div>
  <div
    style="border: 2px solid purple;"
  >
    <p>
      マイタイムライン
    </p>
    <div
      class="react-timeline"
    >
      <div
        class="bg-sidebarBackgroundColor border-b border-borderColor"
        data-testid="headerRootDiv"
        style="display: flex; width: 100%;"
      >
        <div
          data-testid="sidebarHeader"
          style="width: 150px;"
        />
        <div
          class="border border-borderColor overflow-hidden"
          data-testid="headerContainer"
          style="width: calc(100% - 150px);"
        >
          <div
            class="relative"
            data-testid="dateHeader"
            style="height: 30px; width: -450px;"
          >
            <div
              class="absolute flex items-center justify-center h-full border-b cursor-pointer text-sm border-x border-borderColor text-itemColor bg-inherit"
              data-testid="dateHeaderInterval"
              style="left: 52.79146006944444px; width: -2281.25px;"
            >
              <span>
                25
              </span>
            </div>
          </div>
          <div
            class="relative"
            data-testid="dateHeader"
            style="height: 30px; width: -450px;"
          >
            <div
              class="absolute flex items-center justify-center h-full border-b border-l-2 border-borderColor cursor-pointer text-sm bg-dateHeaderBgColor"
              data-testid="dateHeaderInterval"
              style="left: 52.79146006944444px; width: -2281.25px;"
            >
              <span>
                25
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="block overflow-hidden whitespace-nowrap"
        style="height: 60px;"
      >
        <div
          class="overflow-hidden whitespace-nowrap inline-block vertical-top relative"
          style="width: 150px; height: 60px;"
        >
          <div
            style="width: 150px;"
          >
            <div
              class="px-1 overflow-hidden whitespace-nowrap overflow-ellipsis m-0 border-b border-borderColor bg-rowBackgroundEven"
              style="height: 30px; line-height: 30px;"
            >
              group 1
            </div>
            <div
              class="px-1 overflow-hidden whitespace-nowrap overflow-ellipsis m-0 border-b border-borderColor bg-rowBackgroundOdd"
              style="height: 30px; line-height: 30px;"
            >
              group 2
            </div>
          </div>
        </div>
        <div
          class="inline-block whitespace-normal align-top overflow-x-scroll overflow-y-hidden touch-none relative"
          data-testid="scroll-element"
          style="width: -150px; height: 80px; cursor: default;"
        >
          <div
            style="position: absolute; left: 0px; right: 0px; top: 0px; bottom: 0px;"
          >
            <div>
              <div
                class="absolute pointer-events-none top-0 border-l border-borderColor z-[30]"
                style="left: 53.79146006944444px; width: -2281.25px; height: 60px;"
              />
            </div>
            <div
              class="select-none"
            >
              <div
                class="border-b border-borderColor z-[40] bg-rowBackgroundEven"
                style="width: -450px; height: 30px;"
              />
              <div
                class="border-b border-borderColor z-[40] bg-rowBackgroundOdd"
                style="width: -450px; height: 30px;"
              />
            </div>
            <div
              class="rct-items"
            >
              <div
                class="absolute text-xs text-itemColor cursor-pointer bg-itemBackground border border-itemBorderColor z-[80] hover:z-[88]"
                style="left: -174.99999066840277px; top: 35.25px; width: 3px; height: 19.5px; line-height: 19.5px;"
                title="item 1"
              >
                <div
                  class="sticky left-0 overflow-hidden inline-block rounded-sm px-1.5 h-full"
                  style="max-height: 19.5;"
                >
                  item 1
                </div>
              </div>
              <div
                class="absolute text-xs text-itemColor cursor-pointer bg-itemBackground border border-itemBorderColor z-[80] hover:z-[88]"
                style="left: -174.86978233506943px; top: 5.25px; width: 3px; height: 19.5px; line-height: 19.5px;"
                title="item 2"
              >
                <div
                  class="sticky left-0 overflow-hidden inline-block rounded-sm px-1.5 h-full"
                  style="max-height: 19.5;"
                >
                  item 2
                </div>
              </div>
              <div
                class="absolute text-xs text-itemColor cursor-pointer bg-itemBackground border border-itemBorderColor z-[80] hover:z-[88]"
                style="left: -175.52082400173612px; top: 35.25px; width: 3px; height: 19.5px; line-height: 19.5px;"
                title="item 3"
              >
                <div
                  class="sticky left-0 overflow-hidden inline-block rounded-sm px-1.5 h-full"
                  style="max-height: 19.5;"
                >
                  item 3
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
